### PR TITLE
Fix Bazel's compilation from scratch

### PR DESCRIPTION
**Pull request by ochafik**
This makes it possible to build Bazel from sources without bazel, protoc or the grpc-java plugin.

For instance Bazel now builds well in Raspbian (either on a Raspbery Pi 3 or in a Docker container, see [this repo](https://github.com/ochafik/rpi-raspbian-bazel)).

Notes: not sure if Travis is enabled yet for the bazelbuild organization?

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -0,0 +1,19 @@
+services:
+- docker
+script:
+# Ideally we'd just run compile.sh here:
+#
+# - VERBOSE=yes bash ./compile.sh
```
Comment (aehlig) [inline .travis.yml]: I think here is the big misunderstanding of that pull request. Simply calling `bash compile.sh` is not a supported way of building bazel. So, as it is not supposed to work, there is nothing to fix here. The [supported ways of building bazel from source]( https://docs.bazel.build/versions/master/install-compile-source.html) are
- compiling from the distribution artifact, which on top of the sources also has the required generated `protoc` output. (That is used to build a bootstrap bazel that then builds protoc and bazel from sources). This the quickest and easiest way to bootstrap bazel on a new system
- have `protoc` already installed on your system and telling `compile.sh` about its location. This is the way to cooperate with an ambient system providing the needed tools, to leave the way open to eventually build bazel without depending on any third-party tools embedded in our source tree
- have an older version of `bazel` already and `bazel build //src:bazel`; this is the typical workflow for `bazel` developers.

The the [design document](https://www.bazel.build/designs/2016/10/11/distribution-artifact.html) for a background discussion.

Side remark: bazel is currently not using travis.

File: .travis.yml [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+# Unfortunately bazel workers seem to hang in Travis right now so we test the
+# build from scratch in a controlled Docker container (that's... a docker
+# inside a docker inside a VM). This also helps test which packages are actually
+# required for a successful build from scratch (Travis otherwise has tons of
+# installed installed by default).
+#
+# Note: Verbosity is to avoid being killed by Travis after 10min w/o output.
+- echo "set -ex" > docker-test.sh
+- echo "apt-get update" >> docker-test.sh
+- echo "apt-get install -y automake g++ libtool make curl default-jdk git python unzip wget zip" >> docker-test.sh
```
Comment (aehlig) [inline .travis.yml]: ...and here you can another reason for the mentioned design decision: your change is pulling in quite a few new build dependencies (`automake`, `make`) that are not needed, as `protc` can be built with `bazel`.

Comment (ochafik) [reply .travis.yml]: `protoc` can't be built with `bazel` if you don't have `bazel`, which requires `protoc` (*of the correct version*).

**Comment by ochafik** [comment]
(sorry for the long read / thx for your patience!)

Some clarification on my intent: I was trying ([success](https://github.com/ochafik/rpi-raspbian-bazel/releases)) to build Bazel on a Raspberry Pi / Raspbian ('stretch' version), which ships with protoc 3.0.0.

I know [the instructions](https://docs.bazel.build/versions/master/install-compile-source.html) say to use one of the release bundles, but there's none for `armv7l`, so I decided to work off the git repo instead (and to prove the _this step cannot be short cut by checking out the source tree_ assertion wrong).

After witnessing the build errors due to protoc 3.0.0 not handling stuff like `php_namespace` options, my next step was to fetch sources *of the right versions* for both `protoc` + `grpc-java/compiler` and build them, then setup the `PROTOC` & `GRPC_JAVA_PLUGIN` env vars a call ./compile.sh. 

That mostly works, *after*:
- `${OUTPUT_DIR}/src` files are copied / symlinked as `derived`
- Memory used by the bootstrap javac with `-J-Xmx800M` (Rpi blows up with OOM otherwise; might want to make this configurable?)
- PROTOBUF_LIBRARIES array is updated.

My conclusion at this stage was that Bazel build from scratch is broken in 0.8.0 and at HEAD, and that some kind of CI test would be helpful in the end (apologies if I misconcluded, haven't checked what is actually tested). I now realize Bazel has a different CI system than Travis, but there's currently a blind spot / advertised instructions are untested so I did what I could to keep this PR tested.

Then I realized the repo already contained all the sources of the right versions needed to compile both protoc & grpc-java/compiler, making the whole "fetch sources" exercice redundant. I hence set out to compile those sources from the git repo: it's trivial for `grpc-java/compiler`, and for `protoc` nearly everything is there, except for those weird test headers that needed a BUILD edit (but as stated in that file, there are already some Bazel-local edits of that file, so I felt empowered to add more).

Note that this PR only adds an *optional* dependency to autotools, etc: it's still possible to "only" depend on jdk8, protoc + grpc-java/compiler (although to be fair, there's hidden deps like curl, unzip that aren't spelled out in the doc, which only show up when building one a barebones system like that of the debian docker image I used). But *also*, it's now provably possible to build Bazel on a barebones Linux system.

As to where to go from here: if you're not happy merging this PR as a whole, I can definitely split it up, for instance:
- I'll send a separate PR that allows compiling with ./compile.sh + the right PROTOC & GRPC_JAVA_PLUGIN on a desktop:
  (with https://github.com/bazelbuild/bazel/pull/4199/commits/279dfa01acc45a218539dee16336992ef1654eb2, https://github.com/bazelbuild/bazel/pull/4199/commits/723f5edb903577b3764922b96d0d3f5d4d513fab)
- I'll send a modified / configurable version of https://github.com/bazelbuild/bazel/pull/4199/commits/c99bee19358e235b2dc4e06680e65e800df1d185 to allow building on a device with little RAM (like the Rpi3)
- I'll file a bug to have some CI test of the compile from scratch scenario (which could use a modified version of this [defunct script](https://github.com/ochafik/rpi-raspbian-bazel/blob/3042c3c0839dbc222c059da2cecf0e45b6cc10cd/build_from_scratch.sh) that could wget the deps... except it's hard to understand which versions should be pulled, which is why having it done by bootstrap/compile.sh is a superior alternative IMO).

This leaves the issue of excluding test headers from the protoc targets: not sure, should that live in their repo?

Cheers!

File: .travis.yml (hunk 1 above)
Comment (ochafik) [reply .travis.yml]: I actually overlooked the ddoc, thanks for the link!

My proposal is to remove any binaries from the repo / releases (which addresses all of the ddoc's identified shortcomings), and supporting only 2 or 3 ways of building *from sources*:
- with an existing old bazel
- with an existing "correct version" of protoc + grpc-java/compiler
- with without any of these (but with a "standard" opensource / unix build toolchain)

(and incidentally to fix the current scenario 2 - relying on the distro w/ prebuilt protoc - if it's indeed broken as I think it is)

**Comment by aehlig** [comment]
> I know [the instructions](https://docs.bazel.build/versions/master/install-compile-source.html) say to use one of the release bundles, but there's none for `armv7l` [...]

The instructions say to use the distribution artifact---which is _architecture independent_. That's the whole point of it; it contains the sources plus the generated source files (using `protoc`), but nothing architecture specific.

**Comment by ochafik** [comment]
Ah I see, didn't get the difference between -dist and the other artefacts from the release page from my first read, my bad for not RTM :-D. Not sure if it could make sense to state ("(not one of the platform-dependent release archives)" in the doc, or rename -dist to something like -src-with-generated-src or "Source code with generated sources.zip" to match the "Source code" autogenerated artefact? Or "-src-dist" / "Pregenerated bootstrap sources"?

Still, I would find it more standardly-open-source to offer a way for people to build from the git repo or a "real sources" archive on any Unix platform, which this PR allows. Right now the dist file is a shadow way of submitting generated sources, which I would avoid if possible. WDYT?

**Comment by aehlig** [comment]
> Ah I see, didn't get the difference between -dist and the other artefacts
> from the release page [...] Not sure if it could make sense to state ("(not
> one of the platform-dependent release archives)" in the doc

Yes, emphasizing the fact that the distribution archive is architecture independent
might help.

> Still, I would find it more standardly-open-source to offer a way for people to
> build from the git repo or a "real sources" archive on any Unix platform [...]

My understanding is that the proper open-source way would be to not bundle build
dependencies in a project git-repo/"real souces" archive (note that protoc is a
project independent of bazel that is used not only by bazel). Instead people would
bring these build dependencies on their own (in the same way as they bring their
own CC) and that was the idea behing allowing the user to set PROTOC; except that
it is not that easy with PROTOC as there are subtle dependencies on the precise
version used... (so we ended up bundling, and taking the externally set PROTOC
only for bootstrapping, not the proper bazel build).

Concerning the use of generated sources: yes, it is not nice, but on the other
hand not that unusual---most configure scripts are generated nowadays.

So, I do see your point, but I'm a bit hesitant of adding more bootstrap options,
when the true open-source solution would be that PROTOC becomes stable enough to
be just another tools like CC that can simply be used from the host machine.

-- 
Klaus Aehlig
Google Germany GmbH, Erika-Mann-Str. 33, 80636 Muenchen
Registergericht und -nummer: Hamburg, HRB 86891
Sitz der Gesellschaft: Hamburg
Geschaeftsfuehrer: Paul Terence Manicle, Halimah DeLaine Prado

**Comment by mhlopko** [comment]
Can this PR be closed?

**Pull request closed without merging**